src/main/scala/is/hail/utils/BinaryHeap.scala

```diff
@@ -0,0 +1,185 @@
+    if (next > 0)
+      ts(0)
+    else
+      throw emptyHeap()
+
+  def maxPriority(): Long =
+    if (next > 0)
+      ranks(0)
+    else
+      throw emptyHeap()
+
+  def extractMax(): T = {
+    val max = if (next > 0)
+      ts(0)
+    else
+      throw emptyHeap()
+
+    next -= 1
+    m -= max
+    if (next > 0) {
+      put(0, ts(next), ranks(next))
+      bubbleDown(0)
+      maybeShrink()
+    }
+    max
+  }
+
+  def getPriority(t: T): Long =
+    ranks(m(t))
+
+  def decreasePriorityTo(t: T, r: Long) {
+    val i = m(t)
+    assert(ranks(i) > r)
+    ranks(i) = r
+    bubbleDown(i)
+  }
+
+  def increasePriorityTo(t: T, r: Long) {
+    val i = m(t)
+    assert(ranks(i) < r)
+    ranks(i) = r
+    bubbleUp(i)
+  }
+
+  def contains(t: T): Boolean =
+    m.contains(t)
+
+  def toArray: Array[T] = {
+    val trimmed = new Array(size)
+    Array.copy(ts, 0, trimmed, 0, next)
+    trimmed
+  }
+
+  private def parent(i: Int) =
+    if (i == 0) 0 else (i - 1) >>> 1
+
+  private def put(to: Int, t: T, rank: Long) {
+    ts(to) = t
+    ranks(to) = rank
+    m(t) = to
+  }
+
+  private def swap(i: Int, j: Int) {
+    val tempt = ts(i)
+    ts(i) = ts(j)
+    ts(j) = tempt
+    val temprank = ranks(i)
+    ranks(i) = ranks(j)
+    ranks(j) = temprank
+    m(ts(j)) = j
+    m(ts(i)) = i
+  }
+
+  private def maybeGrow() {
+    if (next >= ts.length) {
+      val ts2 = new Array[T](ts.length << 1)
+      val ranks2 = new Array[Long](ts.length << 1)
+      Array.copy(ts, 0, ts2, 0, ts.length)
+      Array.copy(ranks, 0, ranks2, 0, ts.length)
+      ts = ts2
+      ranks = ranks2
+    }
+  }
+
+  private def maybeShrink() {
+    if (next >= minimumCapacity && next < (ts.length >>> 2)) {
+      val ts2 = new Array[T](ts.length >>> 2)
+      val ranks2 = new Array[Long](ts.length >>> 2)
+      Array.copy(ts, 0, ts2, 0, ts2.length)
+      Array.copy(ranks, 0, ranks2, 0, ts2.length)
+      ts = ts2
+      ranks = ranks2
+    }
+  }
+
+  private def bubbleUp(i: Int) {
+    var current = i
+    var p = parent(current)
+    while (ranks(current) > ranks(p)) {
+      swap(current, p)
+      current = p
+      p = parent(current)
+    }
+  }
+
+  private def bubbleDown(i: Int) {
+    var current = i
+    var largest = current
+    var continue = false
+    do {
+      val leftChild = (current << 1) + 1
+      val rightChild = (current << 1) + 2
+
+      if (leftChild < next && ranks(leftChild) > ranks(largest))
+        largest = leftChild
+      if (rightChild < next && ranks(rightChild) > ranks(largest))
+        largest = rightChild
+
+      if (largest != current) {
+        swap(largest, current)
+        current = largest
+        continue = true
+      } else
+        continue = false
+    } while (continue)
+  }
+
+  def checkHeapProperty() {
+    checkHeapProperty(0)
+  }
+
+  private def checkHeapProperty(current: Int) {
+    val leftChild = (current << 1) + 1
+    val rightChild = (current << 1) + 2
+    if (leftChild < next) {
```

Comment:
Sorry, I should have been more clear. Is there a reason you can't just do this:

```
if (leftChild < next) {
assertHeapProperty(leftChild, current)
checkHeapProperty(leftChild)
}

if (rightChild < next) {
assertHeapProperty(rightChild, current)
checkHeapProperty(rightChild)
}
```

Reply:
I have a vague sense that the highest heap property violation is the most interesting one, so I want to `assertHeapProperty` on the `rightChild` before checking the heap property down the left side of the tree.